### PR TITLE
Deposit offer whitelisting

### PR DESCRIPTION
**Pull request by knikos**
## Why this should be merged
This PR addresses the need of maintaining (storing/retrieving) signatures used for enabling certain addresses (whitelisting) to deposit funds on certain deposit offers. 

## How this works
The solution consists of **2 new api methods** (`AddSignature` and `GetSignatures`), a new db table for storing the signatures as well as the service logic that facilitates the storing and retrieving the information.

### API changes
The new deposit offer service consists of two new methods:
 - AddSignatures (POST): requires the fields  `deposit_offer_id`, `addresses` and `signatures`. An optional field `timestamp` can be provided to determine the time period to search for active deposit offers. If none provided, current server timestamp is used.
 - GetSignatures (GET): requires the `address` for which signatures should be retrieved. Furthermore, a timestamp alongside with a signature are required to ensure restriction of access. Finally, a boolean flag called `multisig` is used to determine whether the provided address is a multisig alias or a signlesig address.

### DB changes
The new **db table stores three fields:** `deposit_offer_id`, `address` and `signature`. An index is also created for the whitelisted address as signatures are expected to be queried based on that address.

Notes:

- Adding signatures is also restricted only for valid deposit offer owners to prevent flooding the db with meaningless data.
- Accessing the respective signatures is restricted only for whitelisted addresses or members of whitelisted msig aliases. Each alias member can access individually the signature associated with an alias.
- Multisig: applies to whitelisting and not to ownership. When whitelisting an msig alias, the owner just needs to provide 1 signature as he'll whitelist the alias itself and not every participant.

## How this was tested
Manual and unit tests

### CODE REVIEW

File: dao/deposit_offer_dao.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		db: db,
+	}
+}
+func (d *depositOfferDao) AddSignature(depositOfferID, address, signature string) error {
+	tx, err := d.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	stmt, err := tx.Prepare("INSERT INTO deposit_offer_sigs (deposit_offer_id, address, signature) VALUES (?, ?, ?)")
```
Comment (evlekht) [inline dao/deposit_offer_dao.go]: I think that prepared statements are meant to be reusable, so it makes sense to create them on service-creation, then use them with given args when its needed. So, if you don't want to prepare statements in advance, I'd directly exec sql here

File: dao/deposit_offer_dao.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+	var result []model.DepositOfferSig
+
+	for rows.Next() {
+		var (
+			depositOfferId string
+			signature      string
+		)
+
+		err = rows.Scan(&depositOfferId, &address, &signature)
```
Comment (evlekht) [inline dao/deposit_offer_dao.go]: scanned address would always be equal to arg address, correct?
why scanning it then?

File: service/deposit_offer_service.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+	pub, err := s.secpFactory.RecoverHashPublicKey(signatureArgsHash, signatureBytes)
+	if err != nil {
+		return ids.ShortEmpty, err
+	}
+
+	return pub.Address(), nil
+}
+
+func toBech32Addr(networkID uint32, addr ids.ShortID) (string, error) {
```
Comment (evlekht) [inline service/deposit_offer_service.go]: why not just use `address.Format` ?

File: dao/deposit_offer_dao.go (hunk 1 above)
Comment (knikos) [reply dao/deposit_offer_dao.go]: valid point. I 'll remove stmt preparation.

File: dao/deposit_offer_dao.go (hunk 2 above)
Comment (knikos) [reply dao/deposit_offer_dao.go]: ah you are talking specifically about the address arg. yeah I can spare that...

File: service/deposit_offer_service.go (hunk 3 above)
Comment (knikos) [reply service/deposit_offer_service.go]: lol, missed that. thanks!

File: dao/deposit_offer_dao.go (hunk 1 above)
Comment (evlekht) [reply dao/deposit_offer_dao.go]: I see that stmt became reusable
I'd still suggest to move prepare to `NewDepositOfferDao`

Comment (knikos) [reply dao/deposit_offer_dao.go]: that's right, i failed to update the status in this comment section. Now it's refactored accordingly.

**Review by evlekht — APPROVED** [review]

**Pull request merged**